client/asset/eth/eth.go

```diff
@@ -2402,18 +2536,379 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
+//	   -- If >=10 blocks have passed
+//	     -- Swap status == Initiated
+//		    -- If monitoredTxs contains the tx -> [resubmit redemption, generate new tx with current account nonce]
+//	        -- If monitoredTxs does not contain the tx -> [create a new redemption tx for this secret hash]
+//	     -- Swap status == Redeemed -> [confirmation complete]
+//	     -- Swap status == Refunded -> [return asset.ErrSwapRefunded]
+func (w *assetWallet) confirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeWallet *assetWallet) (*asset.ConfirmRedemptionStatus, error) {
+	const (
+		txConfsNeededToConfirm               = 10
+		blocksToWaitBeforeCoinNotFound       = 10
+		blocksToWaitBeforeCheckingIfReplaced = 10
+	)
+
+	if len(coinID) != common.HashLength {
+		return nil, fmt.Errorf("expected coin ID to be a transaction hash, but it has a length of %d",
+			len(coinID))
+	}
+	var txHash common.Hash
+	copy(txHash[:], coinID)
+	contractVer, secretHash, err := dexeth.DecodeContractData(redemption.Spends.Contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode contract data: %w", err)
+	}
+	hdr, err := w.node.bestHeader(w.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get best header: %w", err)
+	}
+	currentTip := hdr.Number.Int64()
+
+	confStatus := func(confs uint64, txHash common.Hash) *asset.ConfirmRedemptionStatus {
+		return &asset.ConfirmRedemptionStatus{
+			Confs:  confs,
+			Req:    txConfsNeededToConfirm,
+			CoinID: txHash[:],
+		}
+	}
+
+	resubmitTx := func(tx *types.Transaction, nonceOverride *uint64) (*asset.ConfirmRedemptionStatus, error) {
+		replacementTxHash, err := w.resubmitRedemption(tx, contractVer, nonceOverride, feeWallet)
+		if err != nil {
+			return nil, err
+		}
+
+		return confStatus(0, *replacementTxHash), nil
+	}
+
+	// Generally we rely on the transaction retrieved from geth, but we
+	// use our stored monitored tx for two reasons:
+	// - If the transaction was resubmitted, calls to ConfirmRedemption using
+	//   the old tx id will use the monitoredTxs map to determine the new one.
+	// - When geth cannot find the transaction, we can resubmit the redemption
+	//   using the tx stored in the monitoredTxs map.
+	w.monitoredTxsMtx.Lock()
+	monitoredTx, err := w.getLatestMonitoredTx(txHash)
+	if err != nil {
+		// We were unable to find the monitored tx, but we will attempt to continue
+		w.log.Error(err)
+	}
+	var blocksSinceSubmission int64
+	if monitoredTx != nil {
+		monitoredTx.mtx.Lock()
+		defer monitoredTx.mtx.Unlock()
+		if currentTip > monitoredTx.blockSubmitted {
+			blocksSinceSubmission = currentTip - monitoredTx.blockSubmitted
+		}
+		monitoredTxHash := monitoredTx.tx.Hash()
+		if monitoredTxHash != txHash {
+			w.log.Debugf("tx %s was replaced by %s since the last attempt to confirm redemption", txHash, monitoredTxHash)
+			txHash = monitoredTxHash
+		}
+	} else {
+		w.log.Warnf("tx %s is missing from monitored txs", txHash)
+	}
+	w.monitoredTxsMtx.Unlock()
```

Comment:
Why is this held until  here instead of just for `w.getLatestMonitoredTx` (or even inside of that method)?